### ISSUE TITLE
improve Training 

### ISSUE THREAD

**Issue by vishnukv64** [ultralytics/yolov5#205]
Hey, i just want to know what all can i do to increase the accuracy of the model that im training for... i even used the --evolve tag to find out certain hyper parameters that i can fine tune for training!! 

Other than that what else can i do to increase the training accuraacy (mAP)

thanks!!

**Comment by glenn-jocher** [comment]
Your standard options are:
- Use a larger model
- Collect more data
- Evolve hyperparameters (if you have hardware and time)

You can also start your own research or apply other works to the existing architecture and training framework, anything's a possibility there.

**Comment by josephofiowa** [comment]
@vishnukv64 Can you provide more details about the objects you're attempting to detect and their general environment?

Glenn's suggestions are all correct in general. Perhaps there are more specific recommendations based on your context. e.g. if you have very small objects but very large image resolutions, perhaps preprocessing your data into smaller pieces may help, but this is highly dependent on your problem and use case.

**Comment by vishnukv64** [comment]
@glenn-jocher  Im using the Yolov5x model , so @josephofiowa  data preprocessing and data augmentations may do the trick?
my data is not huge it has 3k images for training and the objects are fairly big and visible, but there are lot of objects close to each other ( like u can see multiple stacks of bbox  overlapping each other)

**Comment by glenn-jocher** [comment]
@vishnukv64 well, there's a lot of settings you can tweak, but without seeing your results.png, and a few example batches (i.e. train_batch0.jpg) it's hard to give guidance.

Note that you can also change the testing settings (i.e. iou_threshold, conf_threshold, etc.) to change your mAP. If your objects have a high degree of overlap this may affect your results.

**Comment by vishnukv64** [comment]
![train](https://user-images.githubusercontent.com/33051467/85851800-00557280-b7cd-11ea-8389-e1a61234db96.PNG)
i tried training the model again with these parameters:
'iou_t': 0.50,  # iou training threshold
'fl_gamma': 1.5,  # focal loss gamma (efficientDet default is gamma=1.5)

these are the only changes i made
i wanted to add adam (0.001 but first i thought let me try sgd and see the results and then later on try adam)
and i get some exceptions when its trying to do validation (image)

Note : As the number of epochs increases the GPU memory also increases!!!

**Comment by glenn-jocher** [comment]
@vishnukv64 sure, I see training appears to be proceeding with no errors in your screenshot, and mAP is increasing rather rapidly.

We generally outperform EfficientDet with the current settings. If focal loss would have improved our results we would have implemented it by default, but it does not, as it's intended for detection architectures which combine objectness with classification, and thus suffer from much greater native class imbalance than we do in YOLO.

The messages simply indicate that more memory is being used during NMS. This is due to an increased amount of detections (false positives), caused by your hyperparameter selection. As in all things, if this works for you then great, if not the game goes on and you can try some other settings.

**Comment by vishnukv64** [comment]
@glenn-jocher fl gamma didnt improve the training!! as u said
So now maybe ill try more preprocessing of images and loading it with the given hyperparameters!!
maybe try out some data augmentation methods and train my model.

**Comment by glenn-jocher** [comment]
@vishnukv64 yes, in my experience focal loss leads to excess FPs and lower final mAP in YOLO architectures. The added recall can improve mAP initially, but steady state results underperform.

It's possible it may work for some datasets, but if so I have not seen such a dataset yet. Perhaps extremely sparsely populated datasets might benefit more from focal loss.
